Feat/add password hashing with bcrypt (also fix create user function)

Paulijuz opened the pull request:
lol this was not hard at all

I added the [bcrypt](https://www.npmjs.com/package/bcrypt) package for password hashing. It's one of the older packages for password hashing, but it still remains one of the most popular ones. From what I can see it's considered safe and it's the one that the next js tutorial uses so I think it's good enough for our purposes. 

Alternatively we could use [argon2](https://www.npmjs.com/package/argon2) or [scrypt](https://www.npmjs.com/package/scrypt), but these packages have way fewer downloads. If someone has any objection tho it's not hard to switch (literally only three lines need to be changed.).

JohanHjelsethStorstad reviewed (changes requested):
The only thing is we might want to add Pepper. The default Becrypt algorithm is more than secure enough.

JohanHjelsethStorstad commented on src/auth/password.ts:
```diff
@@ -0,0 +1,24 @@
+import bcrypt from 'bcrypt'
+
+const saltRounds = 12
```
This is a seguestion really.

Paulijuz replied:
i thought about it, but didn't really do anything since we don't have a convention for this case since it's not in the server folder where we use the ConfigVar.js file. i feel like having a ConfigVar.js file in the auth folder might be a bit messy because all the files are so different and not that related. 

I also don't really know if it should be an enviroment variable since it's not something we are going to change often

Paulijuz replied:
no, wait, i changed my mind, i'll add a pepper so i might as well add salt rounds as an enviroment variable. that way the seeding (not that it's important) can use the same amount of salt rounds

JohanHjelsethStorstad commented:
The most important thing is the use of bcrypt.compare to avoid timing attacks.

Paulijuz commented on package.json:
```diff
@@ -18,7 +18,9 @@
     "@fortawesome/free-solid-svg-icons": "^6.5.2",
     "@fortawesome/react-fontawesome": "^0.2.2",
     "@prisma/client": "^5.15.0",
+    "@types/bcrypt": "^5.0.2",
```
whoopsee, fixed

Paulijuz replied:
when it comes to @types/heic-convert it's on main, which is weird, but i removed it

Paulijuz commented on src/prisma/prismaservice/package.json:
```diff
@@ -25,8 +25,10 @@
     "schema": "../schema"
   },
   "devDependencies": {
+    "@types/bcrypt": "^5.0.2",
```
simply forgot to add them as dev dependencies in the main package config, they were correctly added in the package config of prismaserivce

Paulijuz commented:
That's what's being used

Paulijuz commented:
> ERROR: failed to authorize: failed to fetch oauth token: unexpected status from POST request to https://auth.docker.io/token: 503 Service Unavailable

Hmmm

theodorklauritzen commented:
Great work! I only have one question about peppering. After some research at Google, I found this Stackoverlofw post: https://stackoverflow.com/questions/16891729/best-practices-salting-peppering-passwords, where a user discourages the use of peppering. (We should be careful taking his advice since he is only a StackOverflow user.) I think pepper can be nice, but not how it is implemented now. If the pepper value gets exposed we cannot change the stored passwords without asking every user to update their password. A better solution here is to encrypt the password hash (without pepper and still using salt of course). Then if the encryption key is exposed but not the passwords we can decrypt and encrypt all the passwordhashes with a new key. I think all secrets should be easy to change at short notice, in case we are hacked. Here is a nice image I found for illustration:
![image](https://github.com/vevcom/projectNext/assets/19690242/1103186f-f93c-418a-b53b-04b071245f39)

It is worth noticing that Google google doesn't mention password peppering in their document about modern password security. https://cloud.google.com/solutions/modern-password-security-for-system-designers?ref=alphasec.io

Paulijuz commented:
Totally agree @theodorklauritzen. I should have done more research. Although, I was a bit confused by what you meant by "I think pepper can be nice, _but not how it is implemented now_".  The way I understand your argument we should not be using peppering at all. So I removed password peppering and replaced it with encryption.

There are a few things I'm unsure of and would like suggestions/feedback on:
- How should we handle seeding of password? Currently I have simply copy pasted the relevant functions into the seeder. When it was just a few lines it might have been fine, but now with encryption it's way too long to be maintainable. The obvious solution is to share the relevant files between the services, but sharing random files between src to prisma service seems messy.
- Where should the configuration for password hashing and encryption be? Currently only `PASSWORD_SALT_ROUNDS` and `PASSWORD_ENCRYPTION_KEY` are environment variables and the rest are at the top of `password.ts`. I don't feel we're realistically going to have the need to change the rest of the configuration, for example encryption encoding (hex or base64) or encryption algorithm, as part of deployment so they don't merit being environment variables. For non environment variable configuration we have `ConfigVars.ts`, but this would mean having a single file in the auth folder for all the different parts of auth.

theodorklauritzen reviewed (dismissed):
I maent to implement pepper as encryption, but I should have just called it encryption. Everything seems good now :))

Paulijuz commented:
Thanks @theodorklauritzen. <3 I'll also wait @JohanHjelsethStorstad's review as this is quite critical functionality.

JohanHjelsethStorstad reviewed (approved):
It looks good now I think. I don't think having the config at the top of password.ts is a big problem

JohanHjelsethStorstad commented:
I don think having the config in password.ts is a big deal

JohanHjelsethStorstad commented:
We probably won't have a prismaservice in a few months
